### ISSUE TITLE
`ClientConnection`/`ServerConnection` shouldn't hold onto its `config`

### ISSUE THREAD

**Issue by briansmith** [rustls/rustls#665]
The `ClientConfig` structure is pretty large in memory. Except for `key_log`, AFAICT none of the fields are needed beyond the handshake phase of the connection. In an application switches configurations while some connections are still alive, then the old connection config will live until the last connection using it is closed, instead of only living until that connection's handshake is complete. This is a large difference in lifetime.

Also, holding this config during the entire connection lifetime means that our state machines aren't as effective at preventing errors as they should be. For example, in our state machine, we shouldn't have full access to `session_persistence` throughout the handshake. Indeed, `emit_initial_client_hello` should be able to read from the cache; after that, the only thing we should be able to do to the cache is update/insert entries in it; we shouldn't be able to query it. We could do that using an `Arc<dyn WriteOnlyCache>` type field if we refactor the code  to get rid of the full-featured `session_persistence` field. that If we were to make such a change, then we'd eliminate most or all potential for race conditions in Rustls; at least, that appears to be the case based on what we've seen so far.

The above is written for the client side, but the same argument applies to the server side.

**Comment by briansmith** [comment]
More concretely, here's the refactoring I'm suggesting:
1. Remove the `config` fields from each `Connection` type.
2. Add the `config` to `InitialState`.
3. In `InitialState::emit_initial_client_hello`, copy whatever parts of the `config` that are still needed into the `ExpectServerHelloOrHelloRetryRequest` structure. In the case of the session cache, only a insert-or-update-only interface to it should be saved.
4. Throughout the handshake, continue forwarding the parts of the config that are still relevant to the next state, dropping stuff that's no longer needed along the way.

**Comment by djc** [comment]
Oh, nice one. I'd be happy to take care of moving the config parts into the handshake states, though I'd prefer to do it after #660 lands.

**Comment by djc** [comment]
Should be fine.
